contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,949 @@
+        address account
+    ) external whenNotPaused notInBlackList {
+        uint256 _length = operatorAddresses.length;
+        address[] memory stakeCredits = new address[](_length);
+        address credit;
+        for (uint256 i = 0; i < _length; ++i) {
+            if (!_validatorSet.contains(operatorAddresses[i])) revert ValidatorNotExist(); // should never happen
+            credit = _validators[operatorAddresses[i]].creditContract;
+            stakeCredits[i] = credit;
+        }
+
+        IGovToken(GOV_TOKEN_ADDR).syncBatch(stakeCredits, account);
+    }
+
+    /*----------------- system functions -----------------*/
+    /**
+     * @dev This function will be called by consensus engine. So it should never revert.
+     */
+    function distributeReward(address consensusAddress) external payable onlyValidatorContract {
+        address operatorAddress = _consensusToOperator[consensusAddress];
+        Validator memory valInfo = _validators[operatorAddress];
+        if (valInfo.creditContract == address(0) || valInfo.jailed) {
+            emit RewardDistributeFailed(operatorAddress, "INVALID_VALIDATOR");
+            return;
+        }
+
+        uint256 dayIndex = block.timestamp / 1 days;
+        _rewardRecords[operatorAddress][dayIndex] = msg.value;
+        _totalPooledBNBRecords[operatorAddress][dayIndex] = IStakeCredit(valInfo.creditContract).totalPooledBNB();
+
+        IStakeCredit(valInfo.creditContract).distributeReward{ value: msg.value }(valInfo.commission.rate);
+        emit RewardDistributed(operatorAddress, msg.value);
+    }
+
+    /**
+     * @dev Downtime slash. Only the `SlashIndicator` contract can call this function.
+     */
+    function downtimeSlash(address consensusAddress) external onlySlash {
+        address operatorAddress = _consensusToOperator[consensusAddress];
+        if (!_validatorSet.contains(operatorAddress)) revert ValidatorNotExist(); // should never happen
+        Validator storage valInfo = _validators[operatorAddress];
+
+        // slash
+        uint256 slashAmount = IStakeCredit(valInfo.creditContract).slash(downtimeSlashAmount);
+        uint256 jailUntil = block.timestamp + downtimeJailTime;
+        _jailValidator(valInfo, jailUntil);
+
+        emit ValidatorSlashed(operatorAddress, jailUntil, slashAmount, SlashType.DownTime);
+
+        IGovToken(GOV_TOKEN_ADDR).sync(valInfo.creditContract, operatorAddress);
+    }
+
+    /**
+     * @dev Malicious vote slash. Only the `SlashIndicator` contract can call this function.
+     */
+    function maliciousVoteSlash(bytes calldata _voteAddr) external onlySlash {
+        address operatorAddress = _voteToOperator[_voteAddr];
+        if (!_validatorSet.contains(operatorAddress)) revert ValidatorNotExist(); // should never happen
+        Validator storage valInfo = _validators[operatorAddress];
+
+        uint256 dayIndex = block.timestamp / 1 days;
+        if (_felonyMap[dayIndex] >= felonyPerDay) revert NoMoreFelonyToday();
```

Comment:
Please add comment in the code: `This is to prevent many honest validators being slashed at the same time because of implementation bugs`

Reply:
Done